p2p/protocols/protocol_test.go

```diff
@@ -372,11 +372,163 @@ func TestNoHook(t *testing.T) {
 		return nil
 	}
 
-	if err := peer.handleIncoming(handler); err != nil {
+	if err := peer.receive(handler); err != nil {
 		t.Fatal(err)
 	}
 }
 
+func TestPeer_Receive(t *testing.T) {
+	t.Run("OK", func(t *testing.T) {
+		rw := &dummyRW{}
+		peer := NewPeer(nil, rw, createTestSpec())
+		rw.msg = &struct {
+			Content string
+		}{
+			"test content",
+		}
+
+		handler := func(ctx context.Context, msg interface{}) error {
+			if msg.(*perBytesMsgReceiverPays).Content != "test content" {
+				t.Fatal("rw.msg in handler was wrong")
+			}
+			return nil
+		}
+
+		if err := peer.receive(handler); err != nil {
+			t.Fatal(err)
+		}
+	})
+
+	t.Run("ERROR - invalid message", func(t *testing.T) {
+		rw := &dummyRW{}
+		peer := NewPeer(nil, rw, createTestSpec())
+		rw.msg = &struct {
+			Content   string
+			TestField string
+		}{
+			"test content",
+			"test field",
+		}
+
+		handler := func(ctx context.Context, msg interface{}) error {
+			t.Fatal("should not enter here")
+			return nil
+		}
+
+		if err := peer.receive(handler); err != nil {
+			if err.Error() != "Invalid message (RLP error): <= msg #0 (0 bytes): rlp: input list has too many elements for protocols.perBytesMsgReceiverPays" {
+				t.Fatal("error returned from handler is not good")
+			}
+		}
+	})
+
+	t.Run("ERROR - handler error", func(t *testing.T) {
+		rw := &dummyRW{}
+		peer := NewPeer(nil, rw, createTestSpec())
+		rw.msg = &struct {
+			Content string
+		}{
+			"test content",
+		}
+
+		handler := func(ctx context.Context, msg interface{}) error {
+			if msg.(*perBytesMsgReceiverPays).Content != "test content" {
+				t.Fatal("rw.msg in handler was wrong")
+			}
+			return errors.New("test error")
+		}
+
+		if err := peer.receive(handler); err != nil {
+			if err.Error() != errorf(ErrHandler, "(msg code %v): %v", 0, errors.New("test error")).Error() {
+				t.Fatal("error returned from handler is not good")
+			}
+		}
+	})
+}
+
+func TestPeer_Run(t *testing.T) {
+	t.Run("OK", func(t *testing.T) {
+		rw := &dummyRW{}
+		peer := NewPeer(nil, rw, createTestSpec())
+		rw.msg = &struct {
+			Content string
+		}{
+			"test content",
+		}
+
+		handler := func(ctx context.Context, msg interface{}) error {
+			if msg.(*perBytesMsgReceiverPays).Content != "test content" {
+				t.Fatal("rw.msg in handler was wrong")
+			}
+			return nil
+		}
+
+		var wg sync.WaitGroup
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			if err := peer.Run(handler); err != nil {
+				t.Fatal(err)
+			}
+		}()
+
+		time.Sleep(300 * time.Millisecond)
```

Comment:
Do not use plain time.Sleep for synchronisation. Consider using available concurrency tools or retry functions as the last option. This goes for all new tests in this PR.

Reply:
Sounds good. I added it while debugging, these particular tests can be without it at all.